clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/ingestion/heartbeat/HeartbeatMonitoringService.java

```diff
@@ -280,6 +284,72 @@ public void recordFollowerHeartbeat(
     recordHeartbeat(store, version, partition, region, timestamp, followerHeartbeatTimeStamps, isReadyToServe, true);
   }
 
+  /**
+   * Update lag monitor for a given resource replica based on different heartbeat lag monitor action.
+   */
+  public void updateLagMonitor(
+      String resourceName,
+      int partitionId,
+      HeartbeatLagMonitorAction heartbeatLagMonitorAction) {
+    try {
+      String storeName = Version.parseStoreFromKafkaTopicName(resourceName);
+      int storeVersion = Version.parseVersionFromKafkaTopicName(resourceName);
+      Pair<Store, Version> res =
+          getMetadataRepository().waitVersion(storeName, storeVersion, getMaxWaitForVersionInfo(), 200);
+      Store store = res.getFirst();
+      Version version = res.getSecond();
+      if (store == null) {
+        getLogger().error(
+            "Failed to get store for resource: {} with trigger: {}. Will not update lag monitor.",
+            Utils.getReplicaId(resourceName, partitionId),
+            heartbeatLagMonitorAction.getTrigger());
+        return;
+      }
+      if (version == null) {
+        if (!HeartbeatLagMonitorAction.REMOVE_MONITOR.equals(heartbeatLagMonitorAction)) {
+          getLogger().error(
+              "Failed to get version for resource: {} with trigger: {}. Will not update lag monitor.",
+              Utils.getReplicaId(resourceName, partitionId),
+              heartbeatLagMonitorAction.getTrigger());
+          return;
+        }
+        // During version deletion, the version will be deleted from ZK prior to servers perform resource deletion.
+        // It's valid to have null version when trying to remove lag monitor for the deleted resource.
+        version = new VersionImpl(storeName, storeVersion, "");
+      }
+      switch (heartbeatLagMonitorAction) {
+        case SET_LEADER_MONITOR:
+          addLeaderLagMonitor(version, partitionId);
+          break;
+        case SET_FOLLOWER_MONITOR:
+          addFollowerLagMonitor(version, partitionId);
+          break;
+        case REMOVE_MONITOR:
+          removeLagMonitor(version, partitionId);
+          break;
+        default:
+      }
+    } catch (Exception e) {
+      getLogger().error(
+          "Failed to update lag monitor for replica: {} with trigger: {}",
+          Utils.getReplicaId(resourceName, partitionId),
+          heartbeatLagMonitorAction.getTrigger(),
+          e);
+    }
+  }
+
+  ReadOnlyStoreRepository getMetadataRepository() {
+    return metadataRepository;
+  }
+
+  Duration getMaxWaitForVersionInfo() {
+    return maxWaitForVersionInfo;
+  }
+
+  static Logger getLogger() {
```

Comment:
I'm not sure that I've seen this pattern before in our code.  Usually we just call LOGGER, is there a reason for this?

Reply:
Let me check...so I remember I somehow had to do this due to unit test mocking, but maybe it is not true. I can check other unit tests and come back on this.

Reply:
Fixed